### PR TITLE
Quote service

### PR DESCRIPTION
**Pull request by cpage-pivotal**
Consolidates to a single endpoint, `/quotes?q={symbols}`, for looking up stock quotes.

Successfully parses the Yahoo quote responses whether looking up a single quote or multiple quotes.

For all invocations of the quote service, single or multiple, the Yahoo service is the first option. Markitondemand is the fallback option if Yahoo fails.

Xignite support has been removed. Xignite trials only last for 7 days or 250 requests, which is not conducive to a reliable demo environment.

### CODE REVIEW

File: springboottrades-quotes/src/main/java/io/pivotal/quotes/controller/QuoteV1Controller.java [hunk 1]
```diff
@@ -1,137 +0,0 @@
...
- * Rest Controller providing the REST API for the Quote Service. Provides two
- * calls (both HTTP GET methods): - /quote/{symbol} - Retrieves the current
- * quote for a given symbol. - /company/{name} - Retrieves a list of company
- * information for companies that match the {name}.
- * 
- * @author David Ferreira Pinto
- *
- */
-@RestController
-@RequestMapping(value = "/v1")
```
Comment (dpinto-pivotal) [inline springboottrades-quotes/src/main/java/io/pivotal/quotes/controller/QuoteV1Controller.java]: why remove the v1 from the path? its a good thing to have as we can then have multiple versions of the API running.

File: springboottrades-quotes/src/main/java/io/pivotal/quotes/domain/YahooQuoteResponse.java [hunk 2]
```diff
@@ -3,28 +3,25 @@
...
 import com.fasterxml.jackson.annotation.JsonProperty;
 
-
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class YahooQuoteResponse {
 
 	@JsonProperty("query")
-	private YahooResults results;
-	
+	private YahooResult result;
```
Comment (dpinto-pivotal) [inline springboottrades-quotes/src/main/java/io/pivotal/quotes/domain/YahooQuoteResponse.java]: whats the point of this?

File: springboottrades-quotes/src/main/java/io/pivotal/quotes/service/QuoteService.java [hunk 3]
```diff
@@ -36,165 +25,119 @@
...
 	private RestTemplate restTemplate = new RestTemplate();
 
+    /**
+     * Retrieve one or more quotes.
+     *
+     * @param symbols comma delimited list of symbols.
+     * @return a list of quotes.
+     */
+    @HystrixCommand(fallbackMethod = "getMarkitondemandQuotes")
+    public List<Quote> getQuotes(String symbols) throws SymbolNotFoundException {
```
Comment (dpinto-pivotal) [inline springboottrades-quotes/src/main/java/io/pivotal/quotes/service/QuoteService.java]: Isn't this work already done? I feel like i've seen this before

File: springboottrades-quotes/src/main/java/io/pivotal/quotes/service/QuoteService.java [hunk 4]
```diff
@@ -36,165 +25,119 @@
...
+        logger.debug("retrieving quotes for: " + symbols);
+        if ( symbols.isEmpty() ) return new ArrayList<>();
+
+        List<YahooQuote> yahooQuotes;
+        Date createDate;
+        if ( symbols.indexOf( ',' ) == -1 ) {
```
Comment (dpinto-pivotal) [inline springboottrades-quotes/src/main/java/io/pivotal/quotes/service/QuoteService.java]: There is a property to pass to Jackson reader that will read single elements as arrays. no need for tis If statement

File: springboottrades-quotes/src/main/java/io/pivotal/quotes/service/QuoteService.java [hunk 5]
```diff
@@ -36,165 +25,119 @@
...
-		}
-
-		return quote;
+	@HystrixCommand(fallbackMethod = "getQuotesFallback")
+    @SuppressWarnings("unused")
+	public List<Quote> getMarkitondemandQuotes(String symbols) throws SymbolNotFoundException {
+        List<Quote> result = new ArrayList<>();
+        String[] splitSymbols = symbols.split(",");
+
+        for (String symbol : splitSymbols) {
```
Comment (dpinto-pivotal) [inline springboottrades-quotes/src/main/java/io/pivotal/quotes/service/QuoteService.java]: How will this work with Hystrix timeouts in upstreams clients?

**Comment by dpinto-pivotal** [comment]
Hi Corby,

Thanks for the PR. 

Isn't some of this work done already? I feel like I've done some of this already. A few comments:

- we're getting the quotes from Yahoo, but where are we getting the companies list from? still markit? This won't solve the issues we have been having.

- There is a parameter to pass in to the jackson reader to treat single element responses as one element arrays. I'm pretty sure I've addressed this before. Saves the if/then/else statement.

- I want to keep the API version number in the path, as that will allow multiple version running at the same time, ie keep "v1".

**Comment by cpage-pivotal** [comment]
I'm working off the current version of the development branch. If you have
done this already, it's not in the dev branch.

Fixing the markit companies API is outside the scope of this pull request.

I will update the PR to use Jackson's ACCEPT_SINGLE_VALUE_AS_ARRAY
parameter.

It seems like versioning services ('/v1') should be done on all services or
none of them. I'm not completely convinced that the scope of this project
necessitates supporting multiple versioned instances of each service.


On Thu, Jan 14, 2016 at 1:48 PM, David Ferreira Pinto <
notifications@github.com> wrote:

> Hi Corby,
>
> Thanks for the PR.
>
> Isn't some of this work done already? I feel like I've done some of this
> already. A few comments:
>
>    -
>
>    we're getting the quotes from Yahoo, but where are we getting the
>    companies list from? still markit? This won't solve the issues we have been
>    having.
>    -
>
>    There is a parameter to pass in to the jackson reader to treat single
>    element responses as one element arrays. I'm pretty sure I've addressed
>    this before. Saves the if/then/else statement.
>    -
>
>    I want to keep the API version number in the path, as that will allow
>    multiple version running at the same time, ie keep "v1".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dpinto-pivotal/cf-SpringBootTrader/pull/51#issuecomment-171758337>
> .
>



-- 
Corby Page
Senior Field Engineer, Pivotal
713.584.3168

**Pull request merged**